### PR TITLE
feat: allow to topup the wallet via Circle Card/Payment API

### LINKED ISSUE #80: Create Integration with Circle Payment Processing API
# Bounty

- create a page for the user to onramp from Fiat 

## Scope

- integrate Payment Processing API to enable User to Onramp
- save card ID in backend database, associated with the address 
- verify CVV before each Payment
- add callback for XDAI faucet

## Deliverables

- updated code

## Bounty Owner/Gardener

<@your github username> as **<Role>**

## Roles
bounty gardener: @Clearwood / 50 DAI 
bounty worker: @Clearwood  / share
bounty reviewer: name / share

## Gardener checklist

- [x] I have attached one of the `size-` labels to this bounty
- [x] I have submitted this bounty via http://bounty.leapdao.org/viewform
- [x] I have added this bounty to bounty [board](https://github.com/leapdao/mad-hatter/projects/2)

### PR DESCRIPTION
**Pull request by troggy**
Please remember the [Contributing Guidelines](https://github.com/leapdao/meta/blob/master/CONTRIBUTION.md) :heart:

TODO:
- [x] get Circle pub key
- [x] encrypt card data with PGP pub key
- [x] POST a new card request to Circle API, store card Id in local storage for now. Keep in mind that direct call to Circle API wil lbe replaced by our proxy API later on
- [x] implement payment screen (ask for amount, show fee information)
- [x] POST a new payment request to Circle API (will be proxied later as well)
- [x] get saved card from Circle API if card id is present
- [x] retrieve user email from Circle API
- [x] ask for CVC to use saved card (try reusing the payment screen for a new card)
- [x] after payment trigger faucet + show some acknowledgement on screen
- [x] integrate with `payment-backend` lambda
- [x] make the form look better + use basic validation (mandatory fields + format). Use React Final Form or something like this
- [x] redirect to /topup from payment screen if not enough balance
- [ ] handle "Forbidden: payment b2f99144-8741-474c-9887-2424bc90c790 not yet finalized." error from Circle
- [ ] better error messages
...

## Links to relevant issues or information
Closes #80

### CODE REVIEW

**Comment by troggy** [comment]
Circle gotchas:
- idempotency key should be UUID format (otherwise Circle returns obscure error)
- sessionId should be unique for user session and of 0 to 50 length (using `ripemd160(account)` to satisfy that)
- expiry year have to be YYYY format

File: packages/burner-plugin/package.json [hunk 1]
```diff
@@ -17,7 +17,15 @@
     "@burner-wallet/assets": "^1.0.0",
     "@burner-wallet/core": "^1.0.3",
     "@burner-wallet/modern-ui": "^1.0.0",
-    "@whiterabbitjs/client": "^1.3.0"
+    "@types/uuid": "^8.0.0",
```
Comment (mwaa) [inline packages/burner-plugin/package.json:20]: Should be in dev dependencies

**Comment by troggy** [comment]
> is this work flow intentional for testing?

no, it is not intentional. I think it is because of the way "Back" button implemented (not well tested). Is the same happens if you refresh the page?

**Comment by troggy** [comment]
there are still things to fix in this PR :( It is still quite raw. Maybe worth forking out a separate task for corner cases as I've overspend the time for this bounty already

**Comment by mwaa** [comment]
> > is this work flow intentional for testing?
> 
> no, it is not intentional. I think it is because of the way "Back" button implemented (not well tested). Does the same happen if you refresh the page?

Same still happens on refresh

**Comment by mwaa** [comment]
@troggy did a hard refresh workflow works fine as shown on the demo. Feel free to merge.
Maybe spec out the error handling to a different ticket if time on this is over
- Failure on adding card/billing info
- Failure on payment of card
- Failure of faucet payment repolling before asking user for payment again
- Additional ability to change card if payment failed

Might need corresponding ticket on `payment-backend`

**Comment by troggy** [comment]
forked out #93 and #94

**Pull request merged**